buildSrc/src/main/java/com/r3/csde/VNodesHelper.java

```diff
@@ -0,0 +1,288 @@
+package com.r3.csde;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.r3.csde.dtos.*;
+import kong.unirest.HttpResponse;
+import kong.unirest.JsonNode;
+import kong.unirest.Unirest;
+import java.io.FileInputStream;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import static java.net.HttpURLConnection.HTTP_OK;
+
+/**
+ * The VNodesHelper class is used to create and register the Vnodes specified in the static-network-config.json file.
+ */
+
+public class VNodesHelper {
+    private ProjectContext pc;
+    private ProjectUtils utils;
+    private NetworkConfig config;
+    private ObjectMapper mapper;
+
+    public VNodesHelper(ProjectContext _pc, NetworkConfig _config) {
+        pc = _pc;
+        utils = new ProjectUtils(pc);
+        config = _config;
+        Unirest.config().verifySsl(false);
+        mapper = new ObjectMapper();
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    }
+
+    /**
+     * Entry point for setting up vnodes, called from the 5-vNodeSetup task in csde.gradle
+     */
+    public void vNodesSetup() throws CsdeException {
+        List<VNode> requiredNodes = config.getVNodes();
+        createVNodes(requiredNodes);
+        registerVNodes(requiredNodes);
+    }
+
+    /**
+     * Creates vnodes specified in the config if they don't already exist.
+     * @param requiredNodes represents the list of VNodes as specified in the network Config json file (static-network-config.json)
+     */
+    private void createVNodes(List<VNode> requiredNodes) throws CsdeException {
+
+        // Get existing Nodes.
+        List<VirtualNodeInfoDTO> existingVNodes = getExistingNodes();
+
+        // Check if each required vnode already exist, if not create it.
+        for (VNode vn : requiredNodes) {
+            // Match on x500 and cpi name
+            List<VirtualNodeInfoDTO> matches = existingVNodes.stream().filter(existing ->
+                    existing.getHoldingIdentity().getX500Name().equals( vn.getX500Name()) &&
+                    existing.getCpiIdentifier().getCpiName().equals(vn.getCpi()))
+                    .collect(Collectors.toList());
```

Comment:
```suggestion
            List<VirtualNodeInfoDTO> matches = existingVNodes
                .stream()
                .filter(existing ->
                    existing.getHoldingIdentity().getX500Name().equals(vn.getX500Name()) &&
                    existing.getCpiIdentifier().getCpiName().equals(vn.getCpi()))
                .collect(Collectors.toList());
```
this is a cascading way of writing - IMHO much more easier to read

Reply:
changed